<a href="https://colab.research.google.com/github/alexcoy06/Data-Science/blob/main/Project%204/notebooks/Data_Science.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## 1 Project Overview and Setup: Understanding the Purpose

The primary objective of this project is to leverage a comprehensive dataset extracted from the [Glassdoor website](https://www.kaggle.com/datasets/fahadrehman07/data-science-jobs-and-salary-glassdoor), which includes detailed information on data science jobs and salaries. This dataset provides a rich source of information, covering job titles, estimated salaries, job descriptions, company ratings, and essential company details such as location, size, and industry. By conducting a thorough analysis of this dataset, we aim to gain valuable insights into the job market, identify key trends, and understand what to expect when job hunting or researching career opportunities in the data science field.

To make these insights accessible and actionable, the analysis will be visually represented on a self-made website. This approach ensures that the findings are not only informative but also easily interpretable, catering to a wide audience ranging from job seekers to industry analysts. The goal is to provide a clear and comprehensive overview of job market trends, helping users make informed career decisions based on reliable data.

<u>Note: this file will be converted to a python file using `jupyter nbconvert --to script Data_Science.ipynb` in the terminal.<u>


## 2 Initialization

### 2.1 Add imports

To begin our analysis, we first need to set up our Jupyter Notebook environment and import the necessary libraries.

In [1]:
import pandas as pd 
import plotly.express as px 

- Pandas simplifies data organization, converting messy CSV files into tidy, easy-to-handle formats, simplifying analysis.
- And `plotly.express` is a high-level, easy-to-use interface for creating interactive and complex visualizations in Python using Plotly.

### 2.2 Set up DataFrames

To ensure the Jupyter notebook works in all environments, I will configure the necessary paths, using a `dictionary`.

In [2]:
path = {
    'local': '../datasets/Glassdoor_Salary.csv',
    'online': 'https://raw.githubusercontent.com/alexcoy06/Data-Science/main/Project%204/datasets/Glassdoor_Salary.csv'
}

With all paths set, I will now create a function that selects the correct path regardless of the current environment, using `exception handling`.

In [3]:
def load_csv(file_path):
    try:
        df = pd.read_csv(file_path['local'])
    except FileNotFoundError:
            df = pd.read_csv(file_path['online'])
    return df

Now that `load_csv` is created, I can run the `path` dictionary through it to generate the main dataframe for this project. To store and access this dataframe, I will initialize it and assign it to a variable of my choosing.

In [4]:
ds = load_csv(path)

The variable `ds` is now initialized, abbreviated for Data Science.

## 3 Preparing the Data

To effectively utilize the data frame, it's crucial to inspect it first. Addressing any issues found is necessary to ensure the accuracy and usability of the data.

### 3.1 Initial Inspect

First, we need to examine `ds` to determine the type of information it contains.

In [5]:
ds

Job Title  \
0                                       Data Scientist   
1                            Healthcare Data Scientist   
2                                       Data Scientist   
3                                       Data Scientist   
4                                       Data Scientist   
..                                                 ...   
737           Sr Scientist, Immuno-Oncology - Oncology   
738                               Senior Data Engineer   
739  Project Scientist - Auton Lab, Robotics Institute   
740                               Data Science Manager   
741          Research Scientist – Security and Privacy   

                 Salary Estimate  \
0     $53K-$91K (Glassdoor est.)   
1    $63K-$112K (Glassdoor est.)   
2     $80K-$90K (Glassdoor est.)   
3     $56K-$97K (Glassdoor est.)   
4    $86K-$143K (Glassdoor est.)   
..                           ...   
737  $58K-$111K (Glassdoor est.)   
738  $72K-$133K (Glassdoor est.)   
739   $56K-$91K (Glassdoor est.)   
740  $95K-$160K (Glassdoor est.)   
741  $61K-$126K (Glassdoor est.)   

                                       Job Description  Rating  \
0    Data Scientist\r\nLocation: Albuquerque, NM\r\...     3.8   
1    What You Will Do:\r\n\r\nI. General Summary\r\...     3.4   
2    KnowBe4, Inc. is a high growth information sec...     4.8   
3    *Organization and Job ID**\r\nJob ID: 310709\r...     3.8   
4    Data Scientist\r\nAffinity Solutions / Marketi...     2.9   
..                                                 ...     ...   
737  Site Name: USA - Massachusetts - Cambridge\r\n...     3.9   
738  THE CHALLENGE\r\nEventbrite has a world-class ...     4.4   
739  The Auton Lab at Carnegie Mellon University is...     2.6   
740  Data Science ManagerResponsibilities:\r\n\r\nO...     3.2   
741  Returning Candidate? Log back in to the Career...     3.6   

                                     Company Name         Location  \
0                        Tecolote Research\r\n3.8  Albuquerque, NM   
1    University of Maryland Medical System\r\n3.4    Linthicum, MD   
2                                  KnowBe4\r\n4.8   Clearwater, FL   
3                                     PNNL\r\n3.8     Richland, WA   
4                       Affinity Solutions\r\n2.9     New York, NY   
..                                            ...              ...   
737                                    GSK\r\n3.9    Cambridge, MA   
738                             Eventbrite\r\n4.4    Nashville, TN   
739         Software Engineering Institute\r\n2.6   Pittsburgh, PA   
740                           Numeric, LLC\r\n3.2    Allentown, PA   
741           Riverside Research Institute\r\n3.6  Beavercreek, OH   

                  Headquarters                    Size  Founded  \
0                   Goleta, CA   501 to 1000 employees     1973   
1                Baltimore, MD        10000+ employees     1984   
2               Clearwater, FL   501 to 1000 employees     2010   
3                 Richland, WA  1001 to 5000 employees     1965   
4                 New York, NY     51 to 200 employees     1998   
..                         ...                     ...      ...   
737  Brentford, United Kingdom        10000+ employees     1830   
738          San Francisco, CA  1001 to 5000 employees     2006   
739             Pittsburgh, PA   501 to 1000 employees     1984   
740            Chadds Ford, PA       1 to 50 employees       -1   
741              Arlington, VA   501 to 1000 employees     1967   

          Type of ownership  ... avg_salary  \
0         Company - Private  ...       72.0   
1        Other Organization  ...       87.5   
2         Company - Private  ...       85.0   
3                Government  ...       76.5   
4         Company - Private  ...      114.5   
..                      ...  ...        ...   
737        Company - Public  ...       84.5   
738        Company - Public  ...      102.5   
739    College / University  ...       73.5   
740   

The initial step I'd like to take is to remove the `\n` characters, as they likely signify new lines but serve no purpose in `ds`.

In [6]:
ds = ds.replace(r'\n', ' ', regex=True)
ds

Job Title  \
0                                       Data Scientist   
1                            Healthcare Data Scientist   
2                                       Data Scientist   
3                                       Data Scientist   
4                                       Data Scientist   
..                                                 ...   
737           Sr Scientist, Immuno-Oncology - Oncology   
738                               Senior Data Engineer   
739  Project Scientist - Auton Lab, Robotics Institute   
740                               Data Science Manager   
741          Research Scientist – Security and Privacy   

                 Salary Estimate  \
0     $53K-$91K (Glassdoor est.)   
1    $63K-$112K (Glassdoor est.)   
2     $80K-$90K (Glassdoor est.)   
3     $56K-$97K (Glassdoor est.)   
4    $86K-$143K (Glassdoor est.)   
..                           ...   
737  $58K-$111K (Glassdoor est.)   
738  $72K-$133K (Glassdoor est.)   
739   $56K-$91K (Glassdoor est.)   
740  $95K-$160K (Glassdoor est.)   
741  $61K-$126K (Glassdoor est.)   

                                       Job Description  Rating  \
0    Data Scientist\r Location: Albuquerque, NM\r E...     3.8   
1    What You Will Do:\r \r I. General Summary\r \r...     3.4   
2    KnowBe4, Inc. is a high growth information sec...     4.8   
3    *Organization and Job ID**\r Job ID: 310709\r ...     3.8   
4    Data Scientist\r Affinity Solutions / Marketin...     2.9   
..                                                 ...     ...   
737  Site Name: USA - Massachusetts - Cambridge\r P...     3.9   
738  THE CHALLENGE\r Eventbrite has a world-class d...     4.4   
739  The Auton Lab at Carnegie Mellon University is...     2.6   
740  Data Science ManagerResponsibilities:\r \r Ove...     3.2   
741  Returning Candidate? Log back in to the Career...     3.6   

                                    Company Name         Location  \
0                        Tecolote Research\r 3.8  Albuquerque, NM   
1    University of Maryland Medical System\r 3.4    Linthicum, MD   
2                                  KnowBe4\r 4.8   Clearwater, FL   
3                                     PNNL\r 3.8     Richland, WA   
4                       Affinity Solutions\r 2.9     New York, NY   
..                                           ...              ...   
737                                    GSK\r 3.9    Cambridge, MA   
738                             Eventbrite\r 4.4    Nashville, TN   
739         Software Engineering Institute\r 2.6   Pittsburgh, PA   
740                           Numeric, LLC\r 3.2    Allentown, PA   
741           Riverside Research Institute\r 3.6  Beavercreek, OH   

                  Headquarters                    Size  Founded  \
0                   Goleta, CA   501 to 1000 employees     1973   
1                Baltimore, MD        10000+ employees     1984   
2               Clearwater, FL   501 to 1000 employees     2010   
3                 Richland, WA  1001 to 5000 employees     1965   
4                 New York, NY     51 to 200 employees     1998   
..                         ...                     ...      ...   
737  Brentford, United Kingdom        10000+ employees     1830   
738          San Francisco, CA  1001 to 5000 employees     2006   
739             Pittsburgh, PA   501 to 1000 employees     1984   
740            Chadds Ford, PA       1 to 50 employees       -1   
741              Arlington, VA   501 to 1000 employees     1967   

          Type of ownership  ... avg_salary  \
0         Company - Private  ...       72.0   
1        Other Organization  ...       87.5   
2         Company - Private  ...       85.0   
3                Government  ...       76.5   
4         Company - Private  ...      114.5   
..                      ...  ...        ...   
737        Company - Public  ...       84.5   
738        Company - Public  ...      102.5   
739    College / University  ...       73.5   
740       Company 

Personally, I find it significantly improves the appearance.

### 3.2 Simplify `ds`

As observed, `ds` has 28 rows, not all of which are displayed above. Next, I would like to review all the rows and decide which ones are actually needed.

In [7]:
ds_columns = list(ds.columns)
ds_columns

['Job Title',
 'Salary Estimate',
 'Job Description',
 'Rating',
 'Company Name',
 'Location',
 'Headquarters',
 'Size',
 'Founded',
 'Type of ownership',
 'Industry',
 'Sector',
 'Revenue',
 'Competitors',
 'hourly',
 'employer_provided',
 'min_salary',
 'max_salary',
 'avg_salary',
 'company_txt',
 'job_state',
 'same_state',
 'age',
 'python_yn',
 'R_yn',
 'spark',
 'aws',
 'excel']

I believe the following columns are not needed for this project:

- Salary Estimate
- Job Description
- Location
- Headquarters
- Founded
- Sector
- Revenue
- Competitors
- employer_provided
- company_txt
- age

In [8]:
ds = ds.drop(
    ['Salary Estimate', 'Job Description', 'Location','Headquarters','Founded', 'Sector', 'Revenue', 'Competitors', 'employer_provided', 'company_txt', 'age'], axis=1
)
ds.head()

Job Title  Rating  \
0             Data Scientist     3.8   
1  Healthcare Data Scientist     3.4   
2             Data Scientist     4.8   
3             Data Scientist     3.8   
4             Data Scientist     2.9   

                                  Company Name                    Size  \
0                      Tecolote Research\r 3.8   501 to 1000 employees   
1  University of Maryland Medical System\r 3.4        10000+ employees   
2                                KnowBe4\r 4.8   501 to 1000 employees   
3                                   PNNL\r 3.8  1001 to 5000 employees   
4                     Affinity Solutions\r 2.9     51 to 200 employees   

    Type of ownership                          Industry  hourly  min_salary  \
0   Company - Private               Aerospace & Defense       0          53   
1  Other Organization  Health Care Services & Hospitals       0          63   
2   Company - Private                 Security Services       0          80   
3          Government                            Energy       0          56   
4   Company - Private           Advertising & Marketing       0          86   

   max_salary  avg_salary job_state  same_state  python_yn  R_yn  spark  aws  \
0          91        72.0        NM           0          1     0      0    0   
1         112        87.5        MD           0          1     0      0    0   
2          90        85.0        FL           1          1     0      1    0   
3          97        76.5        WA           1          1     0      0    0   
4         143       114.5        NY           1          1     0      0    0   

   excel  
0      1  
1      0  
2      1  
3      0  
4      1

Now, `ds` contains only the columns I am most interested in. All columns are now displayed without any being hidden.

Note that we removed `Salary Estimate` because it would be redundant to keep it, considering we have `min_salary` and `max_salary`.

### 3.3 Check the Data Types

Now, let's execute the `info()` method to ensure that each column contains a value that enables the column to be utilized effectively.

In [9]:
ds.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 742 entries, 0 to 741
Data columns (total 17 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Job Title          742 non-null    object 
 1   Rating             742 non-null    float64
 2   Company Name       742 non-null    object 
 3   Size               742 non-null    object 
 4   Type of ownership  742 non-null    object 
 5   Industry           742 non-null    object 
 6   hourly             742 non-null    int64  
 7   min_salary         742 non-null    int64  
 8   max_salary         742 non-null    int64  
 9   avg_salary         742 non-null    float64
 10  job_state          742 non-null    object 
 11  same_state         742 non-null    int64  
 12  python_yn          742 non-null    int64  
 13  R_yn               742 non-null    int64  
 14  spark              742 non-null    int64  
 15  aws                742 non-null    int64  
 16  excel              742 non

The columns `hourly`, `same_state`, `python_yn`, `R_yn`, `spark`, `aws`, and `excel` contain numerical data types that are not immediately clear. These values are binary, where `1` signifies `true` and `0` signifies `false`.

Each of these columns needs to be converted to a more descriptive format.

In [10]:
ds['hourly'].replace({1: 'hourly', 0: 'salary'}, inplace=True)

ds['same_state'].replace({1: 'local', 0: 'distant'}, inplace=True)

ds.loc[:, ['python_yn', 'R_yn', 'spark', 'aws', 'excel']] = ds[['python_yn', 'R_yn', 'spark', 'aws', 'excel']].replace({1: "Used", 0: "Unused"})

ds.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 742 entries, 0 to 741
Data columns (total 17 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Job Title          742 non-null    object 
 1   Rating             742 non-null    float64
 2   Company Name       742 non-null    object 
 3   Size               742 non-null    object 
 4   Type of ownership  742 non-null    object 
 5   Industry           742 non-null    object 
 6   hourly             742 non-null    object 
 7   min_salary         742 non-null    int64  
 8   max_salary         742 non-null    int64  
 9   avg_salary         742 non-null    float64
 10  job_state          742 non-null    object 
 11  same_state         742 non-null    object 
 12  python_yn          742 non-null    object 
 13  R_yn               742 non-null    object 
 14  spark              742 non-null    object 
 15  aws                742 non-null    object 
 16  excel              742 non

Now the chart is a bit easier to understand, and the data types better represent each column with more relatable values.

Additionally, `min_salary`, `max_salary`, and `avg_salary` are shown in units of tens instead of thousands.

In [11]:
def unit_fix(df, columns):
    for column in columns:
        df[column] *= 1000
    return df

ds = unit_fix(ds, ['avg_salary', 'max_salary', 'min_salary'])

ds.head()

Job Title  Rating  \
0             Data Scientist     3.8   
1  Healthcare Data Scientist     3.4   
2             Data Scientist     4.8   
3             Data Scientist     3.8   
4             Data Scientist     2.9   

                                  Company Name                    Size  \
0                      Tecolote Research\r 3.8   501 to 1000 employees   
1  University of Maryland Medical System\r 3.4        10000+ employees   
2                                KnowBe4\r 4.8   501 to 1000 employees   
3                                   PNNL\r 3.8  1001 to 5000 employees   
4                     Affinity Solutions\r 2.9     51 to 200 employees   

    Type of ownership                          Industry  hourly  min_salary  \
0   Company - Private               Aerospace & Defense  salary       53000   
1  Other Organization  Health Care Services & Hospitals  salary       63000   
2   Company - Private                 Security Services  salary       80000   
3          Government                            Energy  salary       56000   
4   Company - Private           Advertising & Marketing  salary       86000   

   max_salary  avg_salary job_state same_state python_yn    R_yn   spark  \
0       91000     72000.0        NM    distant      Used  Unused  Unused   
1      112000     87500.0        MD    distant      Used  Unused  Unused   
2       90000     85000.0        FL      local      Used  Unused    Used   
3       97000     76500.0        WA      local      Used  Unused  Unused   
4      143000    114500.0        NY      local      Used  Unused  Unused   

      aws   excel  
0  Unused    Used  
1  Unused  Unused  
2  Unused    Used  
3  Unused  Unused  
4  Unused    Used

### 3.4 Duplicate and Missing values

Now that we have all the necessary columns with corrected values, we can begin checking for missing or duplicated values in `ds`. Once this is completed, we will have ensured the accuracy and usability of the data.

In [12]:
ds_missing = ds.isna().sum()
ds_dupl = ds.duplicated(keep=False)

print(f'There are {ds_dupl.sum()} duplicate values, and {ds_missing.sum()} missing values.')

There are 496 duplicate values, and 0 missing values.


It appears that some rows are duplicated, but there are no missing values. Therefore, I want to sample and check some of the duplicated rows.

In [13]:
ds_dupl_check = ds[ds_dupl].sort_values(by='Company Name')
ds_dupl_check.head(6)

Job Title  Rating                      Company Name  \
423  Data Scientist     4.7                    1904labs\r 4.7   
57   Data Scientist     4.7                    1904labs\r 4.7   
9    Data Scientist     4.6                    <intent>\r 4.6   
112  Data Scientist     4.6                    <intent>\r 4.6   
481    Data Analyst     3.6  AXION Healthcare Solutions\r 3.6   
654    Data Analyst     3.6  AXION Healthcare Solutions\r 3.6   

                    Size  Type of ownership                          Industry  \
423  51 to 200 employees  Company - Private                       IT Services   
57   51 to 200 employees  Company - Private                       IT Services   
9    51 to 200 employees  Company - Private                          Internet   
112  51 to 200 employees  Company - Private                          Internet   
481    1 to 50 employees  Company - Private  Health Care Services & Hospitals   
654    1 to 50 employees  Company - Private  Health Care Services & Hospitals   

     hourly  min_salary  max_salary  avg_salary job_state same_state  \
423  salary      111000      176000    143500.0        MO      local   
57   salary      111000      176000    143500.0        MO      local   
9    salary      120000      160000    140000.0        NY      local   
112  salary      120000      160000    140000.0        NY      local   
481  salary       47000       85000     66000.0        NY      local   
654  salary       47000       85000     66000.0        NY      local   

    python_yn    R_yn   spark     aws   excel  
423      Used  Unused    Used  Unused  Unused  
57       Used  Unused    Used  Unused  Unused  
9        Used  Unused    Used  Unused  Unused  
112      Used  Unused    Used  Unused  Unused  
481    Unused  Unused  Unused  Unused    Used  
654    Unused  Unused  Unused  Unused    Used

It's evident that there are duplicated values that need to be removed.

In [14]:
ds.drop_duplicates(inplace=True)
ds = ds.reset_index()
ds_dupl_2 = ds.duplicated(keep=False)

print(f'There are {ds_dupl_2.sum()} duplicate values.')

There are 0 duplicate values.


Now that we have removed all duplicated values, we can proceed to finally use the data.

### 3.5 Hindsight correction

As the saying goes, "hindsight is 20/20." It's often noticed during data analysis, after processing and organizing, that there are negative one numbers and "-1" strings that distort the data and are logically inaccurate. I'll go back to address this issue earlier in my notebook to ensure more accurate calculations later on.

In [15]:
errors = ((ds == -1) | (ds == '-1')).any(axis=1)
errors_check = ds[errors]
errors_check

index                                          Job Title  Rating  \
39      41                                      Data Engineer     3.6   
164    176  Principal Data Scientist with over 10 years ex...    -1.0   
247    300                               Data Operations Lead    -1.0   
294    370    Research Scientist, Immunology - Cancer Biology    -1.0   
323    404  Senior Scientist, Cell Pharmacology/Assay Deve...    -1.0   
360    456                                     Data Scientist    -1.0   
421    581      Scientist – Cancer Discovery, Molecular Assay    -1.0   
444    641  Principal Research Scientist/Team Lead, Medici...    -1.0   

                Company Name                  Size       Type of ownership  \
39            Persivia\r 3.6     1 to 50 employees       Company - Private   
164        CA-One Tech Cloud   51 to 200 employees       Company - Private   
247                     Muso  201 to 500 employees  Nonprofit Organization   
294               Kronos Bio               Unknown       Company - Private   
323               Kronos Bio               Unknown       Company - Private   
360                     ALIN               Unknown       Company - Private   
421  Monte Rosa Therapeutics                    -1                      -1   
444               Kronos Bio               Unknown       Company - Private   

        Industry  hourly  min_salary  max_salary  avg_salary job_state  \
39            -1  salary       68000      129000     98500.0        MA   
164  IT Services  salary      200000      250000    225000.0        CA   
247           -1  salary       85000       90000     87500.0        CA   
294           -1  salary      100000      140000    120000.0        MA   
323           -1  salary      110000      130000    120000.0        MA   
360           -1  salary       81000      140000    110500.0        NY   
421           -1  salary      100000      135000    117500.0        MA   
444           -1  salary      120000      145000    132500.0        MA   

    same_state python_yn    R_yn   spark     aws   excel  
39     distant    Unused  Unused  Unused  Unused    Used  
164    distant      Used  Unused  Unused    Used    Used  
247      local      Used  Unused  Unused  Unused    Used  
294    distant    Unused  Unused  Unused  Unused    Used  
323    distant    Unused  Unused  Unused  Unused  Unused  
360    distant      Used  Unused  Unused  Unused  Unused  
421    distant    Unused  Unused  Unused  Unused    Used  
444    distant    Unused  Unused  Unused  Unused  Unused

As can be seen, for some reason, certain columns such as `Rating` contain a value of -1, which is logically impossible since ratings should range from 0 to 5.

In [16]:
ds = ds[~errors]

The above code edits `ds` to exclude all rows with erroneous values. With this step completed, we are almost done.

There is one more error concerning `job_state`. Instead of showing the initials of each state, there is an instance where the value is ` Los Angeles`.

In [17]:
ds['job_state'].unique()

array([' NM', ' MD', ' FL', ' WA', ' NY', ' TX', ' CA', ' VA', ' MA',
       ' NJ', ' CO', ' IL', ' KY', ' OR', ' CT', ' MI', ' DC', ' OH',
       ' AL', ' MO', ' PA', ' GA', ' IN', ' LA', ' WI', ' NC', ' AZ',
       ' NE', ' MN', ' Los Angeles', ' UT', ' TN', ' DE', ' ID', ' RI',
       ' IA', ' SC', ' KS'], dtype=object)

As previously stated, a value within the 'job_state' column is not a valid state abbreviation.

In [18]:
ds['job_state'] = ds['job_state'].replace(' Los Angeles', 'CA')
ds['job_state'].unique()

array([' NM', ' MD', ' FL', ' WA', ' NY', ' TX', ' CA', ' VA', ' MA',
       ' NJ', ' CO', ' IL', ' KY', ' OR', ' CT', ' MI', ' DC', ' OH',
       ' AL', ' MO', ' PA', ' GA', ' IN', ' LA', ' WI', ' NC', ' AZ',
       ' NE', ' MN', 'CA', ' UT', ' TN', ' DE', ' ID', ' RI', ' IA',
       ' SC', ' KS'], dtype=object)

With this task finished, we can now effectively analyze the data from GlassDoor.

## 4 Criteria for Analysis

### 4.1 Inspecting `ds` for Analysis

Now, my goal is to group the data into distinct data frames, focusing on various points of interest. Let's check `ds` once more to determine these points.

In [19]:
ds

index                                          Job Title  Rating  \
0        0                                     Data Scientist     3.8   
1        1                          Healthcare Data Scientist     3.4   
2        2                                     Data Scientist     4.8   
3        3                                     Data Scientist     3.8   
4        4                                     Data Scientist     2.9   
..     ...                                                ...     ...   
462    695                      Data Architect / Data Modeler     4.3   
463    700                                     Data Scientist     3.4   
464    716  Associate Scientist / Sr. Associate Scientist,...     4.0   
465    732                    Machine Learning Engineer (NLP)     4.1   
466    735                                      Data Engineer     3.9   

                                    Company Name                    Size  \
0                        Tecolote Research\r 3.8   501 to 1000 employees   
1    University of Maryland Medical System\r 3.4        10000+ employees   
2                                  KnowBe4\r 4.8   501 to 1000 employees   
3                                     PNNL\r 3.8  1001 to 5000 employees   
4                       Affinity Solutions\r 2.9     51 to 200 employees   
..                                           ...                     ...   
462                     Medidata Solutions\r 4.3  1001 to 5000 employees   
463                        DatamanUSA, LLC\r 3.4     51 to 200 employees   
464                                23andMe\r 4.0   501 to 1000 employees   
465                       CK-12 Foundation\r 4.1       1 to 50 employees   
466                              Fivestars\r 3.9    201 to 500 employees   

      Type of ownership                                 Industry  hourly  \
0     Company - Private                      Aerospace & Defense  salary   
1    Other Organization         Health Care Services & Hospitals  salary   
2     Company - Private                        Security Services  salary   
3            Government                                   Energy  salary   
4     Company - Private                  Advertising & Marketing  salary   
..                  ...                                      ...     ...   
462    Company - Public  Enterprise Software & Network Solutions  salary   
463   Company - Private                              IT Services  salary   
464   Company - Private                Biotech & Pharmaceuticals  salary   
465   Company - Private                           K-12 Education  salary   
466   Company - Private                                 Internet  salary   

     min_salary  max_salary  avg_salary job_state same_state python_yn  \
0         53000       91000     72000.0        NM    distant      Used   
1         63000      112000     87500.0        MD    distant      Used   
2         80000       90000     85000.0        FL      local      Used   
3         56000       97000     76500.0        WA      local      Used   
4         86000      143000    114500.0        NY      local      Used   
..          ...         ...         ...       ...        ...       ...   
462       63000      110000     86500.0        NY      local    Unused   
463       65000      113000     89000.0        WA    distant    Unused   
464       59000      125000     92000.0        CA    distant    Unused   
465       80000      142000    111000.0        CA      local      Used   
466       62000      113000     87500.0        CA      local      Used   

       R_yn   spark     aws   excel  
0    Unused  Unused  Unused    Used  
1    Unused  Unused  Unused  Unused  
2    Unused    Used  Unused    Used  
3    Unused  Unused  Unused  Unused  
4    Unused  Unused  Unused    Used  
..      ...     ...     ...     ...  
462  Unused  Unused    Used    Used  
463  Unused  Unused  Unused  Unused  
464  Unused  Unused    Used    Used  
465  Unused  Unused    Used    Us

After reviewing 'ds,' I recommend a detailed analysis of the following: `Type of ownership`, `Industry`, skills used, `Ratings`, and the salary ranges.

The remaining columns, while valuable, are less relevant to my needs. For example:
- I work remotely, so the job's location is irrelevant.
- Job titles can vary for the same position.

### 4.2 Column Inspects

To prevent any unforeseen issues, it is advisable to closely examine each column of interest, excluding two types:
- Columns with binary values
- Columns with numerical values

In [20]:
ds['Type of ownership'].unique()

array(['Company - Private', 'Other Organization', 'Government',
       'Company - Public', 'Hospital', 'Subsidiary or Business Segment',
       'Nonprofit Organization', 'Unknown', 'College / University',
       'School / School District'], dtype=object)

All values for the 'Type of ownership' appear to be straightforward.

In [21]:
ds['Industry'].unique()

array(['Aerospace & Defense', 'Health Care Services & Hospitals',
       'Security Services', 'Energy', 'Advertising & Marketing',
       'Real Estate', 'Banks & Credit Unions', 'Consulting', 'Internet',
       'Other Retail Stores', 'Research & Development',
       'Department, Clothing, & Shoe Stores', 'Biotech & Pharmaceuticals',
       'Motion Picture Production & Distribution',
       'Enterprise Software & Network Solutions', 'Insurance Carriers',
       'Insurance Agencies & Brokerages', 'Logistics & Supply Chain',
       'Telecommunications Services', 'IT Services',
       'Computer Hardware & Software', 'Consumer Products Manufacturing',
       'Industrial Manufacturing', 'Metals Brokers',
       'Financial Transaction Processing', 'Sporting Goods Stores',
       'Staffing & Outsourcing', 'Wholesale', 'Mining',
       'Financial Analytics & Research', 'Federal Agencies',
       'Education Training Services',
       'Transportation Equipment Manufacturing', 'Farm Support Servic

The 'Industry' column contains numerous values, many of which are not of interest. Therefore, it would be most beneficial to focus on industries in which I have the most experience.

In [22]:
# Created a list of the industries im interested in
industries = [
    'Computer Hardware & Software', 'Construction', 'Consulting', 'Consumer Products Manufacturing', 'Financial Analytics & Research', 'Insurance Carriers', 'Other Retail Stores', 'Staffing & Outsourcing', 'Video Games'
]

# set a dataframe to only show the rows that include the `industries`
industry = ds[ds['Industry'].isin(industries)]

The `industries` list includes all the industries I am familiar with, either professionally or through personal interest. Using this list, I created `industry`, which contains only the rows from `ds` that match any values listed in `industries`.

### 4.3 Reformating Skill Usage Statistics

We could use `ds` as it is for the skills used, but that wouldn't be very beneficial. We would either end up with single-row data frames or a data frame with rows showing all the different combinations of skills used. Instead, we only want one dataframe showing each skill used with the associated data.

In [23]:
ds_melted = ds.melt(
    id_vars=["Job Title", "Rating", "Company Name", "Size", "Industry", "hourly", "min_salary", "max_salary", "avg_salary", "job_state", "same_state"],
    value_vars=["python_yn", "R_yn", "spark", "aws", "excel"], var_name="skill",
    value_name="required"
)

Now, `ds` is restructured into `ds_melt` by converting the specified columns into rows and creating two new columns. This expands the DataFrame to better display the skills required for each job listing.

In [24]:
ds_melted = ds_melted[ds_melted['required'] == 'Used']

Now we have `ds_melt`, focusing specifically on the skills being used.

Now we can perform the same process as we did in the previous section, using the `groupby()` function.

In [25]:
# Group by the skills and their usage status
skill_stats = ds_melted.groupby(['skill', 'required']).agg({'avg_salary': 'mean', 'Job Title': 'count'}).round().reset_index()
skill_stats = skill_stats.rename(columns={'Job Title': 'total'})
skill_stats

skill required  avg_salary  total
0       R_yn     Used     70750.0      2
1        aws     Used    109327.0    110
2      excel     Used     98045.0    245
3  python_yn     Used    110053.0    257
4      spark     Used    110950.0    111

Unlike the other points of interest, where we focused on salary, with `skill_stats`, our primary interest is in the usage of each skill.

## 5 Job analysis

### 5.1 `Type of ownership`

First, I will examine the different types of ownership and their relationship with salary.

In [26]:
ownership_box = px.box(
    ds,
    x='Type of ownership',
    y='avg_salary',
    title='Average Salary by Type of Ownership',
    labels={'Type of ownership': 'Type of Ownership', 'avg_salary': 'Average Salary'}
)
ownership_box.show()

Based on the data, most types of ownership can offer salaries below $50,000 per year. However, they are also generally capable of paying over $100,000 per year. Notably, hospitals and unknown ownership types tend to offer the lowest salaries.

In [27]:
unknown = ds[ds['Type of ownership'] == 'Unknown']
unknown

index          Job Title  Rating                   Company Name  \
117    125  Project Scientist     4.0  Alliance Source Testing\r 4.0   

                    Size Type of ownership  \
117  51 to 200 employees           Unknown   

                                 Industry  hourly  min_salary  max_salary  \
117  Architectural & Engineering Services  salary       29000       50000   

     avg_salary job_state same_state python_yn    R_yn   spark     aws   excel  
117     39500.0        AL    distant    Unused  Unused  Unused  Unused  Unused

Curiosity led me to investigate jobs with an 'unknown' ownership type. I was relieved to find that this job was in an industry I have no experience. Despite having a rating of four, this job did not require any data scientist skills, which likely explains the lower pay. Overall, this was an interesting discovery.

### 5.2 `Industry`

The previous section has increased my interest in understanding the range of salary rates for the industries in which I have experience.

In [28]:
industry_box = px.box(
    industry,
    x="Industry",
    y="avg_salary",
    labels={"avg_salary": "Average Salary ($)"},
    title="Average Salary by Industry",
)
industry_box.show()

When comparing the industries I have experience in to the different types of business ownerships, I notice some differences. Among the industries I have experience in, only Construction falls completely below $50,000. In contrast, the industries I'm skilled in tend to average fairly close to or well above $100,000.

In [29]:
industry_rating = px.box(
    industry,
    x='Industry',
    y='Rating',
    title='Rating per Industry',
    labels={'Rating': 'Company Rating'}
)

mean_rating = ds['Rating'].mean()

industry_rating.add_hline(
    y=mean_rating,
    line_dash="dash",
    line_color="red"
)

industry_rating.show()


Despite the previous analysis indicating that many of the jobs I have experience in are high-paying, it's worth noting that the ratings don't necessarily reflect them as the best jobs to have. While I **wouldn't** say these jobs **'aren't worth having,'** a significant portion of them fall below the average. The clear winners here appear to be 'Computer Hardware & Software' and 'Consulting'.

### 5.3 Salary

To better understand my previous statement, I would like to further explore the relationship between salary and ratings for various jobs.

In [30]:
salary_rating = px.scatter(
    ds,
    x='avg_salary',
    y='Rating',
    title='Average Salary vs Company Rating',
    labels={'avg_salary': 'Average Salary', 'Rating': 'Company Rating'}
)

salary_rating.add_hline(
    y=mean_rating,
    line_dash="dash",
    line_color="red"
)

salary_rating.show()


Regarding the data in the above graph, it can be observed that there are above-average ratings at both the high and low ends of the salary spectrum. These ratings are heavily concentrated between $50,000 and $150,000. While this distribution makes it challenging to interpret the graph qualitatively, calculating the correlation can provide a clearer understanding of the relationship.

In [31]:
rating_correlation = (ds['avg_salary'].corr(ds['Rating'])).round(4)
rating_correlation

0.1146

With a correlation coefficient of 0.1146, this indicates a weak positive relationship between the average salary and company rating. The correlation is not strong enough to suggest that higher pay corresponds to better-rated jobs.

### 5.4 Skills

Finally, I would like to determine if there are any additional tech skills that might be worth adding to my skillset.

In [32]:
skill_bar = px.bar(
    skill_stats,
    x='skill',
    y='total',
    color='total',
    hover_name='skill',
    labels={'skill': 'Skill Used', 'total': 'Number of Times Used'},
    title='Skills Used'
)
skill_bar.show()

Based on the analysis, I've identified the following key skills and their relevance to my career development:

- The most used skills are Python and Excel, both of which I am experienced in.
- AWS and Spark are also highly used; I plan to gain a better understanding of Spark and might pursue an AWS certification in the future.
- The programming language R appears to be less valuable to learn at this time, as it is the lowest-paid skill on average.

## 6 Conclusion

After analyzing the data, I've come to understand several key outlooks. Firstly, in terms of skill development, I believe I am on the right track to achieve my career goals. Secondly, it can be seen that a high salary doesn't necessarily equate to a high job rating. From personal experience, job satisfaction is often based on how much you enjoy the work, which would be an interesting topic for further analysis. Additionally, based on the industries I have the most experience with and their average salaries, I should expect to easily make close to $100,000.